scipy/_lib/_bunch.py

```diff
@@ -0,0 +1,146 @@
+
+import sys as _sys
+from keyword import iskeyword as _iskeyword
+
+
+def _validate_names(typename, field_names, extra_field_names):
+    """
+    Ensure that all the given names are valid Python identifiers that
+    do not start with '_'.  Also check that there are no duplicates
+    among field_names + extra_field_names.
+    """
+    for name in [typename] + field_names + extra_field_names:
+        if type(name) is not str:
+            raise TypeError('typename and all field names must be strings')
+        if not name.isidentifier():
+            raise ValueError('typename and all field names must be valid '
+                             f'identifiers: {name!r}')
+        if _iskeyword(name):
+            raise ValueError('typename and all field names cannot be a '
+                             f'keyword: {name!r}')
+
+    seen = set()
+    for name in field_names + extra_field_names:
+        if name.startswith('_'):
+            raise ValueError('Field names cannot start with an underscore: '
+                             f'{name!r}')
+        if name in seen:
+            raise ValueError(f'Duplicate field name: {name!r}')
+        seen.add(name)
+
+
+# Note: This code is adapted from CPython:Lib/collections/__init__.py
+def make_tuple_bunch(typename, field_names, extra_field_names=None,
+                     module=None):
+    if len(field_names) == 0:
+        raise ValueError('field_names must contain at least one name')
+
+    if extra_field_names is None:
+        extra_field_names = []
+    _validate_names(typename, field_names, extra_field_names)
```

Comment:
If `field_names` is a single string, `_validate_names` fails with a message that's not super helpful.
```
>>> Result = make_tuple_bunch('Result', 'ab')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\matth\Desktop\scipy\scipy\_lib\_bunch.py", line 40, in make_tuple_bunch
    _validate_names(typename, field_names, extra_field_names)
  File "C:\Users\matth\Desktop\scipy\scipy\_lib\_bunch.py", line 12, in _validate_names
    for name in [typename] + field_names + extra_field_names:
TypeError: can only concatenate list (not "str") to list
```
Is entering a single string for `field_names` something we want to protect against?

Reply:
I'm not too worried about it.  The API for creating these things is strictly for SciPy developers, and as long as some sort of error occurs, the developer will know that something needs to fixed.    A user will never encounter this situation.

I was going to say "a developer should know from reading the docstring of make_tuple_bunch that the argument must be a list", but then I recalled that there was no docstring. 😮   So I added one in a new commit.

Having said that, if others think we need more validation here, I'm not against adding it.  Friendlier error messages are good for developers, too. :)

Note that the `namedtuple` code from which this is derived has code that allows the argument to be either a list of names or a single string of comma-separated names.  I'm not interested in adding that here; it seems like more code to be maintained and tested that provides no real benefit (except maybe typing a few less characters).